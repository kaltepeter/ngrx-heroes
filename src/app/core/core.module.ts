import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';

import { HttpClientInMemoryWebApiModule } from 'angular-in-memory-web-api';
import { InMemHeroService } from './inMemHeroService';

@NgModule({
  declarations: [],
  imports: [
    CommonModule,
    HttpClientModule,
    HttpClientInMemoryWebApiModule.forRoot(InMemHeroService)
  ],
  exports: []
})
export class CoreModule { }
