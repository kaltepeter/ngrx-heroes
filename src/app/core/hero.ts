export class Hero {
  id: number;
  name: string;
}


// for testing
export const generateHero = (idOverride?: number): Hero => ({
  id: idOverride || (Math.floor(Math.random() * 100) + 1),
  name: 'Test name'
});

export const generateHeroArray = (count = 10): Hero[] =>
  // Overwrite random id generation to prevent duplicate IDs:
  Array.apply(null, Array(count)).map((value, index) => generateHero(index + 1));

export const generateHeroMap = (
  heroArray: Array<Hero> = generateHeroArray()
): { ids: Array<number>, entities: any } => ({
  entities: heroArray.reduce(
    (heroMap, hero) => ({ ...heroMap, [hero.id]: hero }),
    {}
  ),
  ids: heroArray.map(hero => hero.id)
});

