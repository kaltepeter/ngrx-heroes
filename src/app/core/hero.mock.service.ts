import { Observable, of } from 'rxjs';
import { Hero } from './hero';

export class HeroMockService {
  getHeroes(): Observable<Hero[]> {
    return of([]);
  }

  getTopHeroes(): Observable<Hero[]> {
    return of([]);
  }

  getHero(id: number): Observable<Hero> {
    return of(null);
  }

  loadHeroes() {
  }

  loadHero(id: number) {
  }
}
