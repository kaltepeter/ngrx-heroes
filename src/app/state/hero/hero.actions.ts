import { Action } from '@ngrx/store';
import { Hero } from '../../core/hero';

export enum HeroActionTypes {
  SearchAllHeroEntities = '[Hero] Search',
  SearchAllHeroEntitiesSuccess = '[Hero] Search Success',
  SearchAllHeroEntitiesFail = '[Hero] Search Fail',

  LoadHeroById = '[Hero] Load By ID',
  LoadHeroByIdSuccess = '[Hero] Load Success',
  LoadHeroByIdFail = '[Hero] Load Fail',
}

// ========================================= SEARCH

export class SearchAllHeroEntities implements Action {
  readonly type = HeroActionTypes.SearchAllHeroEntities;
}

export class SearchAllHeroEntitiesSuccess implements Action {
  readonly type = HeroActionTypes.SearchAllHeroEntitiesSuccess;
  constructor(public payload: { result: Array<Hero> }) {}
}

export class SearchAllHeroEntitiesFail implements Action {
  readonly type = HeroActionTypes.SearchAllHeroEntitiesFail;
  constructor(public payload: { error: string }) {}
}

// ========================================= LOAD BY ID

export class LoadHeroById implements Action {
  readonly type = HeroActionTypes.LoadHeroById;
  constructor(public payload: { id: number }) {}
}

export class LoadHeroByIdSuccess implements Action {
  readonly type = HeroActionTypes.LoadHeroByIdSuccess;
  constructor(public payload: { result: Hero }) {}
}

export class LoadHeroByIdFail implements Action {
  readonly type = HeroActionTypes.LoadHeroByIdFail;
  constructor(public payload: { error: string }) {}
}

export type HeroActions =
  | SearchAllHeroEntities
  | SearchAllHeroEntitiesSuccess
  | SearchAllHeroEntitiesFail
  | LoadHeroById
  | LoadHeroByIdSuccess
  | LoadHeroByIdFail;
