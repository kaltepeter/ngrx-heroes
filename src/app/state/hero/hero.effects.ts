import { Injectable } from '@angular/core';

import { Observable, of } from 'rxjs';
import {
  exhaustMap,
  map,
  catchError,
  switchMap
} from 'rxjs/operators';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';

import {
  HeroActionTypes,
  SearchAllHeroEntities,
  SearchAllHeroEntitiesSuccess,
  SearchAllHeroEntitiesFail,
  LoadHeroById,
  LoadHeroByIdSuccess,
  LoadHeroByIdFail
} from './hero.actions';
import { Hero } from '../../core/hero';
import { HeroService } from '../../core/hero.service';

@Injectable()
export class HeroEffects {
  @Effect()
  search: Observable<Action> = this.actions$.pipe(
    ofType<SearchAllHeroEntities>(HeroActionTypes.SearchAllHeroEntities),
    exhaustMap(() =>
      this.service.getHeroes().pipe(
        map(
          (entities: Array<Hero>) =>
            new SearchAllHeroEntitiesSuccess({ result: entities })
        ),
        catchError(({ message }) =>
          of(new SearchAllHeroEntitiesFail({ error: message }))
        )
      )
    )
  );

  @Effect()
  loadById: Observable<Action> = this.actions$.pipe(
    ofType<LoadHeroById>(HeroActionTypes.LoadHeroById),
    switchMap(action =>
      this.service.getHero(action.payload.id).pipe(
        map((hero: Hero) => new LoadHeroByIdSuccess({ result: hero })),
        catchError(({ message }) =>
          of(new LoadHeroByIdFail({ error: message }))
        )
      )
    )
  );

  constructor(private actions$: Actions, private service: HeroService) {}
}
