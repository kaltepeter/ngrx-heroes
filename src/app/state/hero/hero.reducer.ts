import { Hero } from '../../core/hero';

import * as heroActions from './hero.actions';

export interface HeroState {
  heroes: Hero[];
  hero: Hero;
}

export const initialState: HeroState = {
  heroes: [],
  hero: null
};

export function heroReducer(
  state = initialState,
  action: heroActions.HeroActions
): HeroState {
  switch (action.type) {
    case heroActions.HeroActionTypes.SearchAllHeroEntitiesSuccess: {
      return {
        hero: null,
        heroes: action.payload.result
      };
    }

    case heroActions.HeroActionTypes.LoadHeroByIdSuccess: {
      return {
        heroes: [],
        hero: action.payload.result
      };
    }

    default: {
      return state;
    }
  }
}
