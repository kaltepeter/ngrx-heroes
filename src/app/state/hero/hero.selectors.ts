import { createFeatureSelector, createSelector } from '@ngrx/store';
import * as fromHero from './hero.reducer';

export const selectHeroState = createFeatureSelector<fromHero.State>(
  fromHero.heroFeatureKey
);
