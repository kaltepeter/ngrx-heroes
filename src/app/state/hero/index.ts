import { createSelector, createFeatureSelector } from '@ngrx/store';
import * as fromHeroState from './hero.reducer';
import {HeroState} from './hero.reducer';
export const getHeroState = createFeatureSelector<HeroState>('hero');

export const {
  selectIds: heroIds,
  selectEntities: entities,
  selectAll: heroes,
  selectTotal: heroCount
} = fromHeroState.adapter.getSelectors(getHeroState);

export const topHeroes = createSelector(
  heroes,
  data => data.slice(1, 5)
);

export const selectedHeroId = createSelector(
  getHeroState,
  fromHeroState.getSelectedId
);
export const selectedHero = createSelector(
  selectedHeroId,
  entities,
  (heroId, heroEntities) => heroId && heroEntities[heroId]
);

export const heroLoading = createSelector(
  getHeroState,
  fromHeroState.getLoading
);

export const heroError = createSelector(
  getHeroState,
  fromHeroState.getError
);
