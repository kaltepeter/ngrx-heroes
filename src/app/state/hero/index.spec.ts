import { topHeroes, heroes, selectedHeroId, selectedHero } from './index';
import { Hero } from '../../core/hero';

// State Factory
const createHero = ({
  id = 0,
  name = ''
} = {}): Hero => ({
  id: id,
  name: name || 'name'
});

const createHeroState = ({
  entities = {
    '1': createHero({ id: 1, name: 'Bob' }),
    '2': createHero({ id: 2, name: 'Sue' }),
    '3': createHero({ id: 3, name: 'Mary' }),
    '4': createHero({ id: 4, name: 'Jim' }),
    '5': createHero({ id: 5, name: 'Jane' })
  },
  ids = ['1', '2', '3', '4', '5'],
  selectedId = 0,
  loading = false,
  error = ''
} = {}) => ({
  hero: {
    ids,
    entities,
    selectedId,
    loading,
    error
  }
});

describe('hero selectors', () => {
  it('should return all heroes', () => {
    const state = createHeroState();
    expect(heroes(state)).toMatchSnapshot();
  });

  it('should return top 4 heroes', () => {
    const state = createHeroState();
    expect(topHeroes(state)).toMatchSnapshot();
  });

  it('should return selectedId', () => {
    const state = createHeroState();
    state.hero.selectedId = 3;
    expect(selectedHeroId(state)).toEqual(3);
  });

  it('should return selected Hero', () => {
    const state = createHeroState();
    state.hero.selectedId = 3;
    expect(selectedHero(state)).toMatchSnapshot();
  });
});
