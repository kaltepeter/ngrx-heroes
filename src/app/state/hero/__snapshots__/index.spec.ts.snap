// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hero selectors should return all heroes 1`] = `
Array [
  Object {
    "id": 1,
    "name": "Bob",
  },
  Object {
    "id": 2,
    "name": "Sue",
  },
  Object {
    "id": 3,
    "name": "Mary",
  },
  Object {
    "id": 4,
    "name": "Jim",
  },
  Object {
    "id": 5,
    "name": "Jane",
  },
]
`;

exports[`hero selectors should return selected Hero 1`] = `
Object {
  "id": 3,
  "name": "Mary",
}
`;

exports[`hero selectors should return top 4 heroes 1`] = `
Array [
  Object {
    "id": 2,
    "name": "Sue",
  },
  Object {
    "id": 3,
    "name": "Mary",
  },
  Object {
    "id": 4,
    "name": "Jim",
  },
  Object {
    "id": 5,
    "name": "Jane",
  },
]
`;
